provisioner/data.js

```diff
@@ -1,30 +1,615 @@
+ * which are going to be created but have not yet been created.  We want to
+ * offset the number of units that we'd be creating by the number of pending
+ * capacity units
+ */
+WorkerType.prototype.determineCapacityChange = function(runningCapacity, pendingCapacity, pending) {
+  assert(typeof runningCapacity === 'number');
+  assert(typeof pendingCapacity === 'number');
+  assert(typeof pending === 'number');
+
+  // We need to know how many total capacity units are extant
+  var totalCapacity = pendingCapacity + runningCapacity;
+
+  // We need to know the current ratio of capacity to pending
+  var percentPending = 1 + pending / totalCapacity;
+
+  var change = 0;
+
+  // We only want to scale anything once we have a pending
+  // percentage that exceeds the scalingRatio
+  if (percentPending > this.scalingRatio) {
+    // But when we do, let's submit enough requests that
+    // we end up in an ideal state if all are fulfilled
+    var ideal = (totalCapacity + pending) / this.scalingRatio;
+    change = ideal - totalCapacity;
+  }
+
+  // We need to offset the number of pending jobs by the
+  // number of units that can't yet start running tasks
+  change = change - pendingCapacity;
+
+  debug('"%s" change needed is %d (runningCapacity %d, pendingCapacity %d, pending tasks %d)',
+        this.workerType, change, runningCapacity, pendingCapacity, pending);
+
+  if (totalCapacity + change > this.maxCapacity) {
+    change = this.maxCapacity - totalCapacity;
+    debug('%s, would exceed max, using %d instead', this.workerType, change);
+  } else if (totalCapacity + change < this.minCapacity) {
+    change = this.minCapacity - totalCapacity;
+    debug('%s wouldn\'t be meet min, using %d instead', this.workerType, change);
+  } 
+
+  return Math.round(change);
+  
+};
+
+
+/**
+ * Select region, instance type and spot bids based on the amount of capacity units needed.
+ * The region is picked randomly to distribute load but in future we could do smart things
+ * like say no region can use more than X% of instances.  We use the utility factor to
+ * determine which instance type to bid on.  Utility factor is a relative performance 
+ * indicator.  If we say that a large is worth 2 and a small is worth 1, we'll bid on
+ * smalls unless the price of a large is less than double that of a small.  Utility factors
+ * are currently hardcoded, but in the future we could do smart things like compare runtimes
+ * of things run on each instance type.  The spot bid is calcuated at the one in the price
+ * history multiplied by 1.3 to give a 30% buffer.
+ */
+WorkerType.prototype.determineSpotBids = function(regions, pricing, runningCapacity, pendingCapacity, pending) {
+  assert(regions);
+  assert(pricing);
+  assert(typeof runningCapacity === 'number');
+  assert(typeof pendingCapacity === 'number');
+  assert(typeof pending === 'number');
+  var that = this;
+  
+  var change = this.determineCapacityChange(runningCapacity, pendingCapacity, pending);
```

Comment:
The change can be negative, in which case we should kill existing spot requests...
It doesn't cost us anything of significance to cancel spot requests. In fact this is all the current provisioner does... It requests/cancels spot requests based on `change` where `change = pendingTasks - pendingNodes`.